pkg/ui/src/views/statements/planView.tsx

```diff
@@ -16,115 +16,274 @@ import _ from "lodash";
+
+  renderArrow() {
+    const node = this.props.node;
+    if (node.attrs && node.attrs.length > 0) {
+      if (this.props.expanded) {
+        // up arrow
+        return <div className="arrow-expanded">&uarr;</div>;
+      }
+      // down arrow
+      return <div className="arrow">&darr;</div>;
+    }
   }
-  const collapsed: { [key: string]: CollapsedPlanAttr } = {};
 
-  attrs.forEach((attr) => {
-    if (attr.key === null || attr.value === null) {
-      return;
+  renderMaybeExpandedDetail() {
+    const node = this.props.node;
+    if (node.attrs && node.attrs.length > 0) {
+      if (this.props.expanded) {
+        return (
+          <div className="nodeAttributes">
+            {node.attrs.map( (attr) => (
+              <div key={attr.key} className="attr">
+                {attr.key}
+                {attr.value &&
+                  <span>={attr.value}</span>
+                }
+              </div>
+            ))}
+          </div>
+        );
+      }
     }
-    const existingAttr = collapsed[attr.key];
-    if (!existingAttr) {
-      collapsed[attr.key] = {
-        key: attr.key,
-        value: [attr.value],
-      };
-      return;
-    } else {
-      collapsed[attr.key].value.push(attr.value);
+  }
+
+  renderClassName(warn: boolean) {
+    const node = this.props.node;
+    if (node.attrs && node.attrs.length > 0 && this.props.expanded) {
+      if (warn) {
+        return "nodeDetails warn expanded";
+      }
+      return "nodeDetails expanded";
+    } else if (warn) {
+      return "nodeDetails warn";
     }
-  });
+    return "nodeDetails";
+  }
 
-  const collapsedAttrs = _.values(collapsed);
-  return _.sortBy(collapsedAttrs, (ca) => (
-    ca.key === "table" ? "table" : "z" + ca.key
-  ));
+  render() {
+    const node = this.props.node;
+    const header = planNodeHeaderProps(node);
+    return (
+      <div
+        className={this.renderClassName(header.warn)}
+        onClick={() => {
+          this.props.toggleExpanded();
+        }}
+      >
+        <PlanNodeDescription
+          header={header}
+        /> {this.renderArrow()}
+        {this.renderMaybeExpandedDetail()}
+      </div>
+    );
+  }
+}
+
+interface IPlanNodeProps {
+  node: IPlanNode;
+}
+
+interface IPlanNodeState {
+  expanded: boolean;
+}
+
+class PlanNode extends React.Component<IPlanNodeProps, IPlanNodeState> {
+  constructor(props: IPlanNodeProps) {
+    super(props);
+    this.state = {
+      expanded: false,
+    };
+
+    // This binding is necessary to make `this` work in the callback
+    this.toggleExpanded = this.toggleExpanded.bind(this);
+  }
+
+  // TODO(celia): toggleExpanded() was originally inside <PlanNodeDetails/>,
+  // but I've moved this here for now, because the outer <li> may need
+  // to know state.expanded in order to finish/fix up the styling of the
+  // node connecting lines. Should move this back into <PlanNodeDetails/>
+  // if <li> ends up not needing to know about expanded.
+  toggleExpanded() {
```

Comment:
One Weird Trick: if you write this as `toggleExpanded = () => {`, you don't have to do the `.bind` in the constructor.

Reply:
ooh - that's helpful - thank you!